File: solutions/_725_split_linked_list_in_parts/Solution.java

package _725_split_linked_list_in_parts;

class Solution {
    public ListNode[] splitListToParts(ListNode head, int k) {
        int len = 0;
        ListNode curr = head;
        while (curr != null) {
            len++;
            curr = curr.next;
        }
        int[] sizes = new int[k];
        for (int i = 0; i < k; i++) {
            sizes[i] = len / k;
        }
        for (int i = 0; i < len % k; i++) {
            sizes[i]++;
        }
        ListNode[] result = new ListNode[k];
        curr = head;
        for (int i = 0; i < k; i++) {
            if (sizes[i] == 0) {
                result[i] = null;
                continue;
            }
            result[i] = curr;
            for (int j = 1; j < sizes[i]; j++) {
                curr = curr.next;
            }
            ListNode next = curr.next;
            curr.next = null;
            curr = next;
        }
        return result;

    }
}

/*
 * https://leetcode.com/problems/split-linked-list-in-parts/
 */

File: solutions/_725_split_linked_list_in_parts/solution.py
from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def splitListToParts(
        self, head: Optional[ListNode], k: int
    ) -> List[Optional[ListNode]]:
        n = 0
        curr = head

        while curr:
            n += 1
            curr = curr.next

        sizes = [(n // k) for _ in range(k)]
        for i in range(n % k):
            sizes[i] += 1

        res = [None] * k
        curr = head
        for i in range(k):
            if sizes[i] == 0:
                continue
            res[i] = curr
            for _ in range(1, sizes[i]):
                curr = curr.next
            nxt = curr.next
            curr.next = None
            curr = nxt

        return res


# https://leetcode.com/problems/split-linked-list-in-parts/
